Add device triggers to support setting turn_on event

elupus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for using device automations to configure the turn on action for device. This way of configuring this make it possible to configure from frontend.

One future improvement would be if the media_player entity could detect that there was an action linked to the event. So it could drop support turn on if missing.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #34384 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

elupus committed:
Add device triggers to support setting turn_on event

elupus committed:
Add turn on event

elupus commented:
This pull is totally separate from #34384, however since device registry won't expose entities without unique identifiers, it's not possible to configure these right now. So one option would be to make these truly device linked, and refer to zones instead of entities. That way it should work without that code.

MartinHjelmare commented on tests/components/arcam_fmj/test_device_trigger.py:
```diff
@@ -0,0 +1,91 @@
+async def test_if_fires_on_turn_on_request(hass, calls, player, state):
+    """Test for turn_on and turn_off triggers firing."""
+    state.get_power.return_value = None
+    player.entity_id = "media_player.arcam_fmj_1"
```
Please set up the integration instead and let it add an entity.

For the future, we should remove the `player` fixture, as we don't want to create entities directly in tests of integrations.

elupus replied:
Yup, will do. Too large to fix that fixture in this pull thou.

MartinHjelmare commented on tests/components/arcam_fmj/test_device_trigger.py:
```diff
@@ -0,0 +1,91 @@
+        },
+    )
+
+    await player.async_turn_on()
```
Please call the turn on service instead.

elupus replied:
I'll fix this up when i fixup the whole test_media_player.py (yes i'm aware I said that once before)

elupus replied:
I need to rewrite those tests and clean them up anyway for the config flow pull, so better to do it once i get rid of the whole turn_on config.

elupus replied:
I corrected the test in this file. will move the olders tests over to the new fixture later

MartinHjelmare reviewed (changes requested):
I like this. @balloob what do you think?

elupus commented:
Before i correct the tests. Would it be preferable to detach this from entities? Ie the triggers get configured on the Zones concept of the device rather than the media player entities?

elupus commented:
I swapped over to zone based events. That way the entity related things are more linked to entity platform, and this trigger is more linked to the device.

elupus commented:
That also mean this is no longer in any way dependent on the config flow pull request and works by it self.

balloob commented on homeassistant/components/arcam_fmj/device_trigger.py:
```diff
@@ -0,0 +1,70 @@
+            if event[CONF_ZONE] == config[CONF_SUBTYPE]:
+                hass.async_run_job(action({"trigger": config}))
+
+        return hass.helpers.dispatcher.async_dispatcher_connect(
+            SIGNAL_TURN_ON, _handle_event
+        )
```
I have mixed feelings about this idea. 

1. Device Automations have been limited to use the automation building blocks. This way we make sure that we don't force device automations on people that don't want to. (one day I even want to allow "conversions")
2. I think it's a great idea because once we implement the idea from the mockup that we suggest to create an automation, it will suggest probably with this one, making it clear from the device info page that it is available.
  ![image](https://user-images.githubusercontent.com/1444314/81619046-8f353980-939d-11ea-9734-87b7e3e56d3d.png)
3. It is a hack. If we ever end up implementing device options, this could be just one of the options. But who knows when that will happen 🤷 and we can convert then.

balloob replied:
1 can be solved by making it an event btw.

MartinHjelmare replied:
We normally don't want entities to fire custom events. Should the event in that case be fired separately from the entity by a device api representing instance?

MartinHjelmare replied:
Should we make a "fake" state change in the entity when turning on, which will spawn an event?

balloob replied:
I think in this case we should make an exception and just fire an event.

MartinHjelmare replied:
What will our guideline be then? When is it ok to fire events from the entity? When is it not ok?

balloob replied:
Only if we need to allow plugging in a turn_on command ?

MartinHjelmare replied:
Ok.

elupus replied:
So back to events now.

elupus committed:
Add unique_id based on config entry

elupus committed:
Adjust tests for addition of uuid

elupus committed:
Supported features are now always same

elupus commented:
So i've reverted back to event, and linked to entity_id. I've added a stop-gap fix for unique_id (using the imported config entry id), so this pull request is standalone. Once the complete config_flow pull is in place, that will transition to a real SSDP identifier.

elupus committed:
Switch to player_setup fixture that actually start platform

MartinHjelmare commented on homeassistant/components/arcam_fmj/const.py:
```diff
@@ -5,6 +5,8 @@
 SIGNAL_CLIENT_STOPPED = "arcam.client_stopped"
 SIGNAL_CLIENT_DATA = "arcam.client_data"
 
+EVENT_TURN_ON = "arcam.turn_on"
```
The prefix should be the integration domain as standard.
```suggestion
EVENT_TURN_ON = "arcam_fmj.turn_on"
```

elupus committed:
Update homeassistant/components/arcam_fmj/const.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Great!

bdraco commented:
CI is failing on `tests/components/arcam_fmj/test_device_trigger.py:52:26: F541 f-string is missing placeholders`

bdraco commented:
> CI is failing on `tests/components/arcam_fmj/test_device_trigger.py:52:26: F541 f-string is missing placeholders`

https://github.com/home-assistant/core/pull/35641
